src/ifcopenshell-python/ifcopenshell/file.py

```diff
@@ -35,6 +35,102 @@
     basestring = (str, bytes)
 
 
+class Transaction:
+    def __init__(self, ifc_file):
+        self.file = ifc_file
+        self.operations = []
+
+    def serialise_entity_instance(self, info):
+        for key, value in info.items():
+            if isinstance(value, entity_instance):
+                info[key] = {"type": "single", "id": value.id()}
+            elif isinstance(value, tuple) and isinstance(value[0], entity_instance):
```

Comment:
I usually do `set(map(type, value)) == {entity_instance}`.

We also need `isinstance(tuple) and set(map(type, value)) == {tuple} and set(map(type, sum(value, ())) == {entity_instance}` for LIST OF LIST OF x

We also have entity_instance.walk() for this, something like:

```
is_instance = lambda v: isinstance(v, entity_instance)
get_id = lambda inst: inst.id()
value = entity_instance.walk(value, is_instance, get_id)
```

This recursively visits the list of list of any other attribute value type and applies get_id() when is_instance() returns true. So something like a recursive combination of filter() and map()

Reply:
Wow walk is amazing! I saw it but had no idea how to use it. I hope the last commit makes sense and fixes this!